### ISSUE TITLE
Software transactional memory proposals

### ISSUE THREAD

**Issue by pchiusano** [unisonweb/unison#1672]
@dolio would like your thoughts on this

## Version 1

This version might be hard to implement if the interpreter is in `IO`. The above interface also technically allows mixing `STM` in with other abilities, even though `atomically` doesn't allow this (I think for good reason - imagine mixing `STM` with nondeterminism...).

```Haskell
builtin type TVar a
builtin ability STM

STM.tvar : a ->{STM} TVar a
STM.set : a -> TVar a ->{STM} ()
STM.read : TVar a ->{STM} a
STM.retry : {STM} x
STM.atomically : '{STM} a ->{IO} a
```

Another alternative is to represent STM computations as values, and use some ability sugar to ultimately provide the same interface as the above. Here's a sketch:

## Version 2 (looks easier to implement)

```Haskell
builtin type TVar a
builtin type STM.Computation a -- corresponds directly to `STM a` type in Haskell

-- these won't generally be used directly
STM.Pure : a -> STM.Computation a
STM.Bind : STM.Computation a -> (a -> STM.Computation b) -> STM.Computation b
STM.NewTVar : a -> STM.Computation (TVar a)
STM.Read : TVar a -> STM.Computation a
STM.Set : a -> TVar a -> STM.Computation ()
STM.Retry : STM.Computation x
STM.Atomically : STM.Computation a ->{IO} a

unique ability STM where
  tvar : a ->{STM} (TVar a)
  set : a -> TVar a -> ()
  read : TVar a -> a
  retry : x

STM.toComputation : Request {STM} a -> STM.Computation a
STM.toComputation = cases
  {a} -> STM.Computation.pure a
  {STM.retry -> _} -> STM.Retry
  {STM.tvar a -> resume} -> Bind (STM.NewTVar a) (tvar -> handle resume tvar with STM.toComputation)
  {STM.set a tvar -> resume} -> Bind (STM.Set a tvar) '(handle !resume with STM.toComputation)
  {STM.read tvar -> resume} -> Bind (STM.Read tvar) (a -> handle resume a with STM.toComputation)

STM.atomically : '{STM} a ->{IO} a
STM.atomically c = STM.Atomically (handle !c with STM.toComputation) 
```

## Version 3

In theory we could implement STM as a pure Unison library, using IORef + CAS primitives. Aside from implementation difficulty, I'd miss Haskell's behavior for detecting deadlock ("indefinitely blocked on TVar").

**Comment by dolio** [comment]
I've been thinking about this, and I think option 1 might be possible to implement. It's true that the runtime is using all sorts of mutable `IO` stuff, but it's actually all thread-local. So, I think it might be okay for `atomically` to sort of checkpoint, and make a temporary copy during the transaction. As long as the copy happens for every trial of the transaction, I think that'd work out all right. The only issue would be if `STM` tries to save work by restarting in the middle of a transaction because it knows some variables didn't change, but as I recall it doesn't do that; I think that's Ed's "version control" thing I'm thinking of.

So, I'm sure it will be tricky, but I think it might be possible.

I'll have to think about the ability 'mixing,' but I think it's no different than `IO`. The builtin ability is effective 'on the bottom', and all other abilities must be resolved somehow to get the right signature. It might seem weird intuitively, because the `STM` things are somehow direct mappings, but it is ultimately like layering a bunch of transformers on top of `STM` in Haskell. How everything is interpreted into final `STM` stuff might be pretty complicated, but I think it has to be well-defined.

**Comment by atacratic** [comment]
Will there be an equivalent of [`unsafeIOToSTM`](http://hackage.haskell.org/package/base-4.14.0.0/docs/GHC-Conc.html#v:unsafeIOToSTM)?  I think there ought to be (one day at least), otherwise some valid stuff becomes impossible: cases where you need to get some external input in order to work out how to complete your STM transaction, but you don't know _what_ input until you're halfway through the transaction.  This seems to be common enough to have earned a place in the [realworldhaskell tutorial on STM](http://book.realworldhaskell.org/read/software-transactional-memory.html), (section 'I/O and STM').  To work around it, you'd have to break your STM transaction up, introducing more intermediate states that would be visible to other threads, and which might conflict with invariants you were hoping to maintain.

**Comment by dolio** [comment]
Is this complete, or is there more to add? I don't think I made `unsafeIOToSTM`, but it should be too hard to add, I think.

**Comment by pchiusano** [comment]
Yeah we can close. I'm not so sure on `unsafeIOToSTM` for Unison but @atacratic you can create a separate issue to track that.
